src/matrix_storage.cpp

```diff
@@ -0,0 +1,973 @@
+
+  // No list, no worries.
+  if (list.size() <= 0)
+    return NULL;
+
+  strlcpy(tmp, name, sizeof(tmpname));
+  if (!(number = get_number(&tmp, sizeof(tmpname))))
+    return NULL;
+
+  for(obj_data* i : list)  {
+    // Invisible to you: Blocked by quest protections.
+    if (!staff_bit && ch_is_blocked_by_quest_protections(ch, i, FALSE, FALSE))
+      continue;
+
+    if (keyword_appears_in_obj(tmp, i)) {
+      if (++j == number)
+        return i;
+    }
+  }
+
+  return NULL;
+}
+
+matrix_file* create_matrix_file(obj_data *storage, int load_origin) {
+  struct matrix_file *new_file = new matrix_file();
+  new_file->idnum = _next_matrix_id();
+  new_file->load_origin = load_origin;
+  new_file->creation_time = time(0);
+
+  if (storage) {
+    new_file->in_obj = storage;
+    new_file->next_file = storage->files;
+    storage->files = new_file;
+  }
+
+  new_file->dirty_bit = TRUE;
+  return new_file;
+}
+
+matrix_file* clone_matrix_file(struct matrix_file *source) {
+  struct matrix_file *copy = create_matrix_file(NULL, source->load_origin);
+
+  if (source->name) copy->name = strdup(source->name);
+  copy->file_type = source->file_type;
+  copy->program_type = source->program_type;
+  copy->rating = source->rating;
+  copy->size = source->size;
+  copy->wound_category = source->wound_category;
+  copy->is_default = source->is_default;
+  copy->creator_idnum = source->creator_idnum;
+  if (source->content) copy->content = strdup(source->content);
+  copy->skill = source->skill;
+  copy->linked = source->linked;
+  copy->compression_factor = source->compression_factor;
+
+  copy->work_phase = source->work_phase;
+  copy->work_ticks_left = source->work_ticks_left;
+  copy->work_original_ticks_left = source->work_original_ticks_left;
+  copy->work_successes = source->work_successes;
+
+  copy->from_host_color = source->from_host_color;
+  copy->from_host_vnum = source->from_host_vnum;
+  
+  copy->last_decay_time = source->last_decay_time;
+  copy->dirty_bit = TRUE;
+
+  return copy;
+}
+
+/* remove a matrix file from an object */
+/* NOTE, THIS WILL ALSO DELETE THE SQL ENTRY */
+void file_from_obj(struct matrix_file * obj, bool do_sql_delete = TRUE)
+{
+  struct matrix_file *temp;
+  struct obj_data *obj_from;
+
+  if (obj->in_obj == NULL)
+  {
+    log("error (handler.c): trying to illegally extract obj from obj");
+    return;
+  }
+  obj_from = obj->in_obj;
+  
+  REMOVE_FROM_LIST(obj, obj_from->files, next_file);
+  if (do_sql_delete) {
+    snprintf(buf, sizeof(buf), "DELETE FROM matrix_files WHERE idnum=%ld; ", 
+      obj->idnum);
+    mysql_wrapper(mysql, buf);
+    obj->dirty_bit = TRUE;
+  }
+
+  obj->in_obj = NULL;
+  obj->next_file = NULL;
+}
+
+/* remove a matrix file from an object */
+void file_from_host(struct matrix_file * obj)
+{
+  struct matrix_file *temp;
+  struct host_data *host_from;
+
+  if (obj->in_host == NULL)
+  {
+    log("error (handler.c): trying to illegally extract obj from obj");
```

Comment:
"obj from host" here. Can this be a `log_vfprintf` and include a loggable_representation for easier debugging?